iceoryx2/src/port/waitset.rs

```diff
@@ -0,0 +1,349 @@
+// Copyright (c) 2023 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! # Example
+//!
+//! ## Use [`AttachmentId::originates_from()`](crate::port::waitset::AttachmentId)
+//!
+//! ```no_run
+//! use iceoryx2::prelude::*;
+//! # use core::time::Duration;
+//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! # let node = NodeBuilder::new().create::<ipc::Service>()?;
+//! # let event_1 = node.service_builder(&"MyEventName_1".try_into()?)
+//! #     .event()
+//! #     .open_or_create()?;
+//! # let event_2 = node.service_builder(&"MyEventName_2".try_into()?)
+//! #     .event()
+//! #     .open_or_create()?;
+//!
+//! let mut listener_1 = event_1.listener_builder().create()?;
+//! let mut listener_2 = event_2.listener_builder().create()?;
+//!
+//! let waitset = WaitSetBuilder::new().create::<ipc::Service>()?;
+//! let _guard_1 = waitset.attach(&listener_1)?;
+//! let _guard_2 = waitset.attach(&listener_2)?;
+//!
+//! while waitset.timed_wait(|attachment_id| {
+//!     let listener = if attachment_id.originates_from(&listener_1) {
+//!         &listener_1
+//!     } else {
+//!         &listener_2
+//!     };
+//!
+//!     while let Ok(Some(event_id)) = listener.try_wait_one() {
+//!         println!("received notification {:?}", event_id);
+//!     }
+//! }, Duration::from_secs(1)) != Ok(WaitEvent::TerminationRequest) {}
```

Comment:
What do you think about this
```rs
let event_handler = |attachment_id| {
    // ...
}

while waitset.timed_wait(event_handler, Duration::from_secs(1)) != Ok(...) {}
```
Else one has already forgotten whats going on when reaching the line with `}, Duration::from...`

I would also propose to switch the parameter and have the timeout as first parameter. It reads more fluid
> timed wait for 'duration' and do event handling if not a termination request

What do you think about a
```rs
waitset.run(event_handler);
```
If someone wants a timeout, it would look like
```rs
let deadline = Deadline(100ms);
waitset.attach(deadline);
```

Reply:
I thought about this, and this gets even more complex. Let's assume you have a `WaitSet` with multiple attachments. These attachments have a liveliness, the first every 23ms, the second every 50ms and the third every 78ms. You want to wake up when no message was received and the timeout was hit. So you require a deadline timer with a priority queue.

But this would not be attachable with the attach API since it would require firing an fd event. This could be handled completely via the internal wait time - you just need to know how long you have to wait and what timeout you hit.

So my idea was to add this later to the builder parameters as a parameter.

I like `WaitSet::run()` - but there are already very simplistic examples where you need a variable deadline. Take a look at this loop: https://github.com/elfenpiff/iceoryx2/blob/iox2-390-event-based-comm-example/examples/rust/event_based_communication/publisher.rs#L38

The goal here is to send out a sample every second. But in between new events may be triggered so you have to rewait on the waitset but now have to subtract the elapsed time.

Reply:
Also, what if you do not want to block ... would it be then `WaitSet::run` and `WaitSet::try_run` ?!

Reply:
I was thinking of something where you setup your application the then just run the event loop. There would be nor `try_run` or `timed_run`. Maybe that's something for a higher-level API and the `WaitSet` needs more fine grained control, just like it is now.